### PR TITLE
Fix on click edit mode states

### LINKED ISSUE #18299: Notebooks: Can't type in code cell after clicking on active code cell
- Azure Data Studio Version: insiders

Steps to Reproduce:
1. Create new notebook
2. Add code cell
3. Click on the active code cell

Expected: Should be able to type in code cell
Actual: Cannot type in code cell unless the cell is clicked again

### PR DESCRIPTION
**Pull request by barbaravaldez**
- Manage the edit mode of the cell when clicking on a different cell, outside the cell and on the same cell.

- Fix smoke tests

This PR is for #18299 and addresses #18366

### CODE REVIEW

File: test/smoke/src/sql/areas/notebook/notebook.test.ts [hunk 1]
```diff
@@ -40,6 +40,7 @@ export function setup(opts: minimist.ParsedArgs) {
 			await app.workbench.sqlNotebook.waitForPlaceholderGone();
 
 			const text1: string = 'SEL';
+			await app.workbench.sqlNotebook.clickOnCodeCell();
```
Comment (lucyzhang929) [inline test/smoke/src/sql/areas/notebook/notebook.test.ts:43]: Are we no longer able to type right away after adding a new code cell?

File: src/sql/workbench/contrib/notebook/browser/notebook.component.ts [hunk 2]
```diff
@@ -307,6 +302,32 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
 		this.model.onScroll.fire();
 	}
 
+	public clickOffCell(e: MouseEvent) {
+		if (this.model.activeCell?.isEditMode) {
+			// when clicking outside of cells toggle edit mode to false
+			this.toggleEditMode();
+		} else {
+			this.unselectActiveCell();
```
Comment (lucyzhang929) [inline src/sql/workbench/contrib/notebook/browser/notebook.component.ts:310]: The current behavior in stable will toggle edit mode and unselect the active cell when you click outside the cell. Do we want to change the behavior to only toggle edit mode?

File: src/sql/workbench/contrib/notebook/browser/notebook.component.ts [hunk 3]
```diff
@@ -307,6 +302,32 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
...
+		}
+	}
+
+	public clickOnCell(cell: ICellModel, event?: MouseEvent) {
+		if (event) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebook.component.ts:315]: `event?.stopPropagation()`

File: src/sql/workbench/contrib/notebook/browser/notebook.component.ts [hunk 4]
```diff
@@ -307,6 +302,32 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
...
+			event.stopPropagation();
+		}
+		if (!this.model.activeCell || this.model.activeCell.id !== cell.id) {
+			if (this.model.activeCell?.isEditMode) {
+				// before activating the new cell, change the edit mode of the current active cell
+				this.toggleEditMode();
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebook.component.ts:321]: It really feels like we should have an editMode setter...all this checking for isEditMode and then calling toggle seems like a recipe for a bug later on. And in most of these cases we don't actually want to toggle it - we want to specifically set it to on or off.

File: test/smoke/src/sql/areas/notebook/notebook.test.ts (hunk 1 above)
Comment (barbaravaldez) [reply test/smoke/src/sql/areas/notebook/notebook.test.ts:43]: We can type right away when creating a new code cell or after a single click

Comment (lucyzhang929) [reply test/smoke/src/sql/areas/notebook/notebook.test.ts:43]: I think @barbaravaldez was saying that we are able to type right away so we shouldn't need this new step in the smoke test right?

File: src/sql/workbench/contrib/notebook/browser/notebook.component.ts (hunk 2 above)
Comment (barbaravaldez) [reply src/sql/workbench/contrib/notebook/browser/notebook.component.ts]: Just reverted to the stable behavior

File: test/smoke/src/sql/areas/notebook/notebook.test.ts (hunk 1 above)
Comment (barbaravaldez) [reply test/smoke/src/sql/areas/notebook/notebook.test.ts]: No change on the current behavior. I thought this was fixing the smoke test failure but it ended up being another issue

File: test/automation/src/sql/notebook.ts [hunk 5]
```diff
@@ -121,6 +121,7 @@ export class Notebook {
 	}
 
 	// Code Cell Actions
+	private static readonly codeCellSelector = 'code-component div.editor';
```
Comment (lucyzhang929) [inline test/automation/src/sql/notebook.ts:124]: We don't need the changes in this file anymore right?

**Review by lucyzhang929 — APPROVED** [review]

File: test/automation/src/sql/notebook.ts (hunk 5 above)
Comment (barbaravaldez) [reply test/automation/src/sql/notebook.ts:124]: oh that's right let me remove this

File: src/sql/workbench/contrib/notebook/browser/notebook.component.ts [hunk 6]
```diff
@@ -312,6 +314,22 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
...
+		event?.stopPropagation();
+		this.unselectActiveCell();
+	}
+
+	public clickOnCell(cell: ICellModel, event?: MouseEvent) {
+		event?.stopPropagation();
+		if (!this.model.activeCell || this.model.activeCell.id !== cell.id) {
+			this.selectCell(cell);
+			if (cell.cellType === CellTypes.Code) {
+				// toggleEditMode on single click for code cells
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/notebook.component.ts:327]: Remove, not really helpful (and incorrect function name)

**Review by Charles-Gagnon — APPROVED** [review]
Kind of hard to follow exactly what all the changes are doing but don't see anything immediately concerning so think this is fine to get in and then do another round of testing to make sure things are looking good.
